test/functional/normalization/flatten_labels_test.go

```diff
@@ -17,30 +17,55 @@ var _ = Describe("[Normalization] Fluentd normalization", func() {
 
 	var (
 		framework *functional.FluentdFunctionalFramework
+		pb        *functional.PipelineBuilder
 	)
 
 	BeforeEach(func() {
 		framework = functional.NewFluentdFunctionalFramework()
-		functional.NewClusterLogForwarderBuilder(framework.Forwarder).
-			FromInput(logging.InputNameApplication).
-			ToFluentForwardOutput()
-		Expect(framework.Deploy()).To(BeNil())
-		Expect(framework.WritesApplicationLogs(1)).To(BeNil())
+		pb = functional.NewClusterLogForwarderBuilder(framework.Forwarder).
+			FromInput(logging.InputNameApplication)
 	})
 	AfterEach(func() {
 		framework.Cleanup()
 	})
 
-	It("should remove 'kubernetes.labels' and create 'kubernetes.flat_labels' with an array of 'kubernetes.labels'", func() {
-		raw, err := framework.ReadApplicationLogsFrom(logging.OutputTypeFluentdForward)
-		Expect(err).To(BeNil(), "Expected no errors reading the logs")
-		logs, err := types.ParseLogs(utils.ToJsonLogs(raw))
-		Expect(err).To(BeNil(), "Expected no errors parsing the logs")
-		//verify the new key exists
-		Expect(logs[0].Kubernetes.FlatLabels).To(Not(BeNil()), fmt.Sprintf("Expected to find the kubernetes.flat_labels key in %#v", logs[0]))
+	Context("for ES output", func() {
+		It("should remove 'kubernetes.labels' and create 'kubernetes.flat_labels' with an array of 'kubernetes.labels'", func() {
+			pb.ToElasticSearchOutput()
+			Expect(framework.Deploy()).To(BeNil())
+			Expect(framework.WritesApplicationLogs(1)).To(BeNil())
+			raw, err := framework.GetLogsFromElasticSearch(logging.OutputTypeElasticsearch, logging.InputNameApplication)
+			Expect(err).To(BeNil(), "Expected no errors reading the logs")
+			Expect(raw).To(Not(BeEmpty()))
 
-		//verify we removed the old key
-		Expect(logs[0].Kubernetes.Labels).To(BeNil(), fmt.Sprintf("Did not expect to find the kubernetes.labels key in %#v", logs[0]))
+			// Parse log line
+			var logs []types.ApplicationLog
+			err = types.StrictlyParseLogs(raw, &logs)
+			Expect(err).To(BeNil(), "Expected no errors parsing the logs")
+
+			//verify the new key exists
+			Expect(logs[0].Kubernetes.FlatLabels).To(Not(BeNil()), fmt.Sprintf("Expected to find the kubernetes.flat_labels key in %#v", logs[0]))
+
+			//verify we removed the old key
+			Expect(logs[0].Kubernetes.Labels).To(BeNil(), fmt.Sprintf("Did not expect to find the kubernetes.labels key in %#v", logs[0]))
+		})
+	})
+	Context("for non-ES output", func() {
+		It("should not remove 'kubernetes.labels' and not add 'kubernetes.flat_labels'", func() {
```

Comment:
@alanconway I am on board with this change since I'm confident any users of non-ES outputs would prefer a map of labels and likely don't use "flat_labels", but this does break backwards compatibility.  I am inclined to go forward with the change and risk a bug to fix in the future if needed

Reply:
As for the legacy syslog and Fluentd method, I am afraid that the flattening labels code are already removed by this commit 2a19585ad75e34b4a62c99c8f1e93d72a0b7c9fa. It depends on the result of this discussion, but I will fix them to keep the backward compatibility.

Reply:
@jcantrill  OK  let's try going with the change and using labels for everything except ES, which gets flat_labels.